funlet-simple-message/functions/funlet-simple-message.js

```diff
@@ -0,0 +1,210 @@
+/*
+  Simple Message Funlet
+
+  Description:
+    Play one or several messages from recordings in MP3 format or through
+    speech-to-text synthesis in a choice of languages and voices.
+
+    This is an upgrade of the Simple Message Twimlet [1].
+    Designed to be backward-compatible with the Twimlet, it was
+    extended to offer better support for internationalization.
+
+  Contents:
+    1. Configuration
+    2. Input Parameters
+    3. Output Helpers
+    4. Main Handler
+    5. Other Exports
+    6. References
+*/
+
+/*
+  1. Configuration
+
+  Here you can change values for the input parameters,
+  directly in the script.
+
+  These values will be superseded by HTTP parameters and properties
+  defined in the environment. You can customize the names and priorities
+  of these various parameters in the next section: Input Parameters.
+*/
+
+let config = {
+  // a list of one or several text string messages,
+  // each being a recording URL or a text to say.
+  messages: [""],
+
+  // language code for conversion of text-to-speech messages,
+  // e.g. 'en' or 'en-gb'
+  language: "en",
+
+  // voice for text-to-speech messages, one of 'man', 'woman' or 'alice'
+  voice: "alice"
+};
+
+/*
+  2. Input Parameters
+
+  Each input parameter Foo is read by a separate function getFoo()
+  which takes one parameter for each source:
+
+    * params - object, the set of HTTP parameters
+               from the URL (GET) or the body (POST) of the query
+    * env - object, the set of environment properties
+            defined in the Twilio account
+    * config - object, the configuration object
+               defined above in this script
+
+  The HTTP parameters are considered first, then environment properties,
+  then the script parameters. This can be customized in the functions below.
+*/
+
+function getMessages(params, env, config) {
```

Comment:
The [twimlet documentation](https://www.twilio.com/labs/twimlets/message) lists the following examples:

```
 http://twimlets.com/message?Message[0]=http://myserver.com/hello.mp3&Message[1]=Thank+You+For+Calling 
```

which returns:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
  <Play>http://myserver.com/hello.mp3</Play>
  <Say>Thank You For Calling</Say>
</Response>
```

and 

```
http://twimlets.com/message?Message[0]=http://myserver.com/1.mp3&Message[1]=http://myserver.com/2.mp3&Message[2]=I+Just+Played+A+File&Message[3]=I+Just+Said+Some+Text
```

which returns:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
  <Play>http://myserver.com/1.mp3</Play>
  <Play>http://myserver.com/2.mp3</Play>
  <Say>I Just Played A File</Say>
  <Say>I Just Said Some Text</Say>
</Response>
```

The difference here is that the twimlet directly references using `Message[0]`, `Message[1]` etc for the query parameters and it supports that version of setting parameters. This function will return an empty response if you try to use `Message[0]` as a paramter.

To be a fully functional drop in for the twimlet, we need to support that format too.

Reply:
Agreed. I will add a test for these cases to check and confirm the issue.

Reply:
@philnash while I am working on this, there is another compatibility issue that I would like to bring to your attention. The URLs in these examples are in a display format (the parameters are not URL-encoded) apart from the spaces in the text messages, which are replaced with a `+`. Now, this is a [PHP convention](https://www.php.net/manual/en/function.urlencode.php) which [I have chosen to disregard](https://github.com/twilio-labs/function-templates/issues/13#issuecomment-524101631), relying on the standard URL-encoding produced by the Twimlet Generator, where space characters are correctly encoded as `%20` instead:

<img width="797" alt="Twimlet Generator URL for Example 1" src="https://user-images.githubusercontent.com/151276/64293671-5d95e400-cf6d-11e9-96ba-9e8e49e02d3a.png">

If you feel like the `+` characters should be treated as space as well, for the sake of compatibility, I can make the change.

Reply:
The `+` encoding of space is a non-issue. The string arrives correctly decoded in the `event` object provided to the handler:

```
{ Message:
   [ 'http://myserver.com/1.mp3',
     'http://myserver.com/2.mp3',
     'I Just Played A File',
     'I Just Said Some Text' ] }
```

It is true, however, that the values are left as separate properties in the `event` object when using numbered parameters `Message[0]`, `Message[1]`... instead of unnumbered parameters of the form `Message[]`:

```
{ 'Message[0]': 'http://myserver.com/hello.mp3',
  'Message[1]': 'Thank You For Calling' }
```

which does result in an empty response in the Simple Message Funlet at the moment:

```
[SIMPLE-MESSAGE-4] Play a greeting then say some text (Example 1)
<?xml version="1.0" encoding="UTF-8"?>
<Response/>

[SIMPLE-MESSAGE-5] Play two greetings then say two text blurbs (Example 2)
<?xml version="1.0" encoding="UTF-8"?>
<Response/>
```

Reply:
Yeah, `+` and `%20` are interchangable and I'm not worried about that.

You're right about how the `Message`s come through when listed like that. I assume the reading of numbered parameters like that is also a PHP thing. It might be a bit of a hassle.

Do you have an idea on how to approach turning those parameters into the messages we expect?

Reply:
### Note to Myself

In twilio-run, the `event` object is [built here](https://github.com/twilio-labs/twilio-run/blob/7bd4f7c6689c52fa36484e31aef716d3584ef0cb/src/runtime/route.ts#L24) merging GET properties [`req.query`](https://expressjs.com/en/api.html#req.query) and POST properties [`req.body`](https://expressjs.com/en/api.html#req.body) decoded by the Express middleware.

Express in turn uses a query string parser, which can be [configured](https://expressjs.com/en/api.html#app.set) through the `query string` property, and defaults to the extended query parser [qs](https://github.com/ljharb/qs). This is where the decoding of parameters and grouping of array values is implemented:

* [Parsing Arrays](https://github.com/ljharb/qs#parsing-arrays)

This should actually handle both cases (`Message[]` and `Message[0]`) in a similar way, except that only numbered indexes from `0` to `20` are considered. This limit can be changed by setting the `arrayLimit` limit parameter to a value different from `20`.

The behavior that we observe would correspond to an `arrayLimit` set to `0`, which is actually preferable for the Simple Menu twimlet. Otherwise, option keys below 20 would be lost, and these menu options would get a sequential number instead:

> When creating arrays with specific indices, qs will compact a sparse array to only the existing values preserving their order

Reply:
> Do you have an idea on how to approach turning those parameters into the messages we expect?

Yes,  I already have part of this logic in place in the Simple Menu Funlet:

```
for( let name of Object.keys(params) ) {
    let matches = /^Options\[([0-9]+)\]$/.exec( name );
    if( matches !== null ) {
      let digits = matches[1];
      options[ digits ] = params[ name ];
    }
  }
```

If both  indexed and non-indexed `Message` parameters are present, I will push the non-indexed ones at the end of the array.

Reply:
Indexed parameters are now correctly handled:

```
[SIMPLE-MESSAGE-4] Play a greeting then say some text (Example 1)
<?xml version="1.0" encoding="UTF-8"?>
<Response>
  <Play>http://myserver.com/hello.mp3</Play>
  <Say language="en" voice="alice">Thank You For Calling</Say>
</Response>

[SIMPLE-MESSAGE-5] Play two greetings then say two text blurbs (Example 2)
<?xml version="1.0" encoding="UTF-8"?>
<Response>
  <Play>http://myserver.com/1.mp3</Play>
  <Play>http://myserver.com/2.mp3</Play>
  <Say language="en" voice="alice">I Just Played A File</Say>
  <Say language="en" voice="alice">I Just Said Some Text</Say>
</Response>
```

I will generalize the change to list parameters in other Funlets.